batchproducer/batchproducer.go

```diff
@@ -233,15 +232,22 @@ func (b *batchProducer) Stop() error {
 }
 
 func (b *batchProducer) setRunning(running bool) {
-	b.runningMu.Lock()
-	defer b.runningMu.Unlock()
-	b.running = running
+	var newValue int32
+	if running {
+		newValue = 1
+	} else {
+		newValue = 0
+	}
+
+	oldValue := b.running
+
+	for swapped := false; !swapped; {
```

Comment:
Do we start-stop all that often to warrant this spinloop? I think that calling start/stop while the component is already started/stopped to be a caller error and thus should yield no effect.

I'd rewrite this as simply:

```go
func (b *batchProducer) setRunning(running bool) bool {
    return atomic.CompareAndSwapInt32(&b.running, 0, 1)
}
```